Token-driven distributed processes (#27)
exactly-once/Home: content/posts/token-driven-processes.md

Token-driven distributed business processes

> Long ago (1987) some industry leaders like Hector Garcia-Molina realized that transactions are not a suitable way of maintaining consistency of long-running processes. Despite this fact, the reliance on transactions for consistency of writes done across multiple processes and long timeframes continued with technologies such as Microsoft Distributed Transaction Coordinator or XA implementing the 2-phase commit protocol. 

Reviewer: ```suggestion
Long ago (1987) some industry leaders like Hector Garcia-Molina realized that transactions are not a suitable way of maintaining consistency of long-running processes. Despite this fact, the reliance on transactions for consistency of writes done across multiple processes over long timeframes continued with technologies such as Microsoft Distributed Transaction Coordinator or XA implementing the 2-phase commit protocol. 
```

The most sophisticated (ab)use of distributed transaction technology was probably the [WS-AtomicTransaction spec](https://specs.xmlsoap.org/ws/2004/10/wsat/wsat1104.pdf) from 2004, which stood out as too enterprisey even compared to the other WS-* specifications.

> The trend seem to have started to change shortly afterwards, around 2007, when Pat Helland published his seminal pager "Life beyond Distributed Transactions: an Apostate's Opinion" where he explained why long-running distributed processes are better realized as a series of message exchanges between stateful entities.

Reviewer: ```suggestion
The trend seems to have started to change shortly afterward, around 2007 when Pat Helland published his seminal paper "Life beyond Distributed Transactions: an Apostate's Opinion" where he explained why long-running distributed processes are better implemented as a series of message exchanges between stateful entities.
```

Around the same time messaging frameworks such as NServiceBus and MassTransit became well-known. They both took advantage of the distributed transaction mechanism available in Windows (MS DTC) to ensure exactly-once message processing across a distributed system. The idea was simple. Distributed transactions are bad because they limit autonomy of software components. By limiting the scope of the transaction to just two parties, the messaging node and a database, one can avoid the pitfals while enjoying the benefits. And the benefits were not trivial because having the guarantee that messages don't get duplicated means the business logic code could focus on the actual business task rather than figuring out if a given message has already been processed.

> Fast forward 15 years and in today's world distributed system are the norm. Long-running business processes are split into multiple steps. Each step may be executed by different computer under control of a different operating system and implemented in a different programming language. Based on that description one could think that the software industry took Pat Hellands advice seriously and used these frameworks to everyone's benefit. Hardly anything could be farther from truth. The prevalent communication style today is HTTP-based web services and devs continue to struggle with unrelible applications that lose and/or duplicate information.

Reviewer: ```suggestion
Fast forward 15 years and in today's world distributed systems are the norm. Long-running business processes are split into multiple steps. Each step may be executed by a different computer under the control of a different operating system and implemented in a different programming language. Based on that description one could think that the software industry took Pat Helland's advice seriously and used these frameworks to everyone's benefit. Hardly anything could be farther from the truth. The prevalent communication style today is HTTP-based web services and devs continue to struggle with unreliable applications that lose and/or duplicate information.
```

> This could be party explained by the fact that reliable messaging frameworks did not receive support from major software vendors which were preocupied with extending their market shares by selling snake oil in form of deduplication built into the infrastructure e.g. Azure ServiceBus or AWS SQS FIFO. In the meantime technology that supported the reliability of these messaging solutions (like DTC) aged pretty fast. Although a viable replacement in the form of the Outbox pattern has been available for 8 years now, only recently it started to gain attention it deserves (quote Oskar).

Reviewer: ```suggestion
This could be partly explained by the fact that reliable messaging frameworks did not receive support from major software vendors who were preoccupied with extending their market shares by selling snake oil in the form of deduplication built into the infrastructure e.g. Azure ServiceBus or AWS SQS FIFO. Meantime technology that was the bedrock of the reliability for many of the messaging solutions (like DTC) aged pretty fast. Although a viable replacement in the form of the Outbox pattern has been available for 8 years now, only recently has it started to gain the attention it deserves (quote Oskar).
```

We hope you can see a pattern now. Business processes span multiple systems. Within these systems the signal travel via messaging. Between the systems the signal travels via HTTP. Messaging and HTTP are just different mediums by which the signal is propagated. We can call the path the signal takes _a conversation_. The conversation is directed graph consisting of activities executed in many different components. In order to ensure the conversation does not corrupt the data we need to ensure that each activity's side effects are applied exactly-once.

> With that introduction you might already realized why we called the idea token-driven distributed business processes. When any of these systems we observe interacts with the extrnal world a new signal can be generated. However, before that signal is emited, a token for processing it is created. Regardless of the medium the signal uses (a message, a http request, a pigeon), processing the token is required to process a signal. Processing a signal consumes the token and if there is no token, the signal is ignored. This means that when a given signal is duplicated anywhere along the way, only one copy is processed by an activity. All other copies are dropped. When that activity processes a signal, it may emit zero or more new signals that continue the same _conversation_. Before these signals are emited, tokens for them are processed in the same way as the inigial token was. 

Reviewer: ```suggestion
With that introduction, you might have already realized why we called the idea token-driven distributed business processes. When any of these systems we observe interacts with the external world a new signal can be generated. However, before that signal is emitted, a token for processing it is created. Regardless of the medium, the signal uses (a message, an HTTP request, a pigeon), processing the token is required to process a signal. Processing a signal consumes the token and if there is no token, the signal is ignored. This means that when a given signal is duplicated anywhere along the way, only one copy is processed by an activity. All other copies are dropped. When that activity processes a signal, it may emit zero or more new signals that continue the same _conversation_. Before these signals are emitted, tokens for them are processed in the same way as the initial token was. 
```

> This approach puts the idea of deduplication on its head. Instead of recording that a component _did process_ a signal after the processing is done, we create a token for processing a signal before it is emitted. This ensures that at any point in time the space occupied by the deduplication data (tokens) is proportional to the amount of in-flight signals. It also provides deterministic, rather than statistical, exacactly-once processing guarantees.

Reviewer: ```suggestion
This approach puts the idea of deduplication on its head. Instead of recording that a component _did process_ a signal after the processing is done, we create a token for processing a signal before it is emitted. This ensures that at any point in time, the space occupied by the deduplication data (tokens) is proportional to the number of in-flight signals. It also provides deterministic, rather than statistical, exactly-once processing guarantees.
```
